Tools/clinic/clinic.py

```diff
@@ -4292,6 +4292,37 @@ def dedent(self, line):
 
 StateKeeper = Callable[[str | None], None]
 
+class ParamState(enum.IntEnum):
+    """Parameter parsing state.
+
+     [ [ a, b, ] c, ] d, e, f=3, [ g, h, [ i ] ]   <- line
+    01   2          3       4    5           6     <- state transitions
+    """
+    # Before we've seen anything.
+    # Legal transitions: to LEFT_SQUARE_BEFORE or REQUIRED
+    START = 0
+
+    # Left square backets before required params.
+    LEFT_SQUARE_BEFORE = 1
+
+    # In a group, before required params.
+    GROUP_BEFORE = 2
+
+    # Required params, positional-or-keyword or positional-only (we
+    # don't know yet). Renumber left groups!
+    REQUIRED = 3
+
+    # Positional-or-keyword or positional-only params that now must have
+    # default values.
+    OPTIONAL = 4
+
+    # In a group, after required params.
+    GROUP_AFTER = 5
+
+    # Right square brackets after required params.
+    RIGHT_SQUARE_AFTER = 6
```

Comment:
It's possible to give enums per-attribute docstrings. You could do it like this:

```diff
--- a/Tools/clinic/clinic.py
+++ b/Tools/clinic/clinic.py
@@ -4,6 +4,7 @@
 # Copyright 2012-2013 by Larry Hastings.
 # Licensed to the PSF under a contributor agreement.
 #
+from __future__ import annotations

 import abc
 import ast
@@ -4298,29 +4299,33 @@ class ParamState(enum.IntEnum):
      [ [ a, b, ] c, ] d, e, f=3, [ g, h, [ i ] ]   <- line
     01   2          3       4    5           6     <- state transitions
     """
-    # Before we've seen anything.
-    # Legal transitions: to LEFT_SQUARE_BEFORE or REQUIRED
-    START = 0

-    # Left square backets before required params.
-    LEFT_SQUARE_BEFORE = 1
+    def __new__(cls, value: int, doc: str) -> ParamState:
+        member = int.__new__(cls, value)
+        member._value_ = value
+        member.__doc__ = doc
+        return member

-    # In a group, before required params.
-    GROUP_BEFORE = 2
+    START = 0, """\
+    State we're in before we've seen anything.

-    # Required params, positional-or-keyword or positional-only (we
-    # don't know yet). Renumber lefft groups!
-    REQUIRED = 3
+    Legal transitions: to LEFT_SQUARE_BEFORE or REQUIRED.
+    """
+
+    LEFT_SQUARE_BEFORE = 1, "Left square backets before required params."
+    GROUP_BEFORE = 2, "In a group, before required params."
+
+    REQUIRED = 3, """\
+    Required params, positional-or-keyword or positional-only (we don't know yet).
+
+    Renumber lefft groups!
+    """

-    # Positional-or-keyword or positional-only params that now must have
-    # default values.
-    OPTIONAL = 4
+    OPTIONAL = 4, "Positional-or-keyword or positional-only params that now must have default values."

-    # In a group, after required params.
-    GROUP_AFTER = 5
+    GROUP_AFTER = 5, "In a group, after required params."

-    # Right square brackets after required params.
-    RIGHT_SQUARE_AFTER = 6
+    RIGHT_SQUARE_AFTER = 6, "Right square brackets after required params."
```

If you do that, then `help()` gives you very pretty interactive documentation:

```pycon
Tools/clinic>python
>>>from clinic import ParamState
>>> help(ParamState)
Help on class ParamState in module clinic:

class ParamState(enum.IntEnum)
 |  ParamState(value, names=None, *, module=None, qualname=None, type=None, start=1, boundary=None)
 |
 |  Parameter parsing state.
 |
 |   [ [ a, b, ] c, ] d, e, f=3, [ g, h, [ i ] ]   <- line
 |  01   2          3       4    5           6     <- state transitions
 |
 |  Method resolution order:
 |      ParamState
 |      enum.IntEnum
 |      builtins.int
 |      enum.ReprEnum
 |      enum.Enum
 |      builtins.object
 |
 |  Methods defined here:
 |
 |  __format__(self, format_spec, /)
 |      Default object formatter.
 |
 |  __new__(cls, value)
 |      Create and return a new object.  See help(type) for accurate signature.
 |
 |  ----------------------------------------------------------------------
 |  Data and other attributes defined here:
 |
 |  GROUP_AFTER = <ParamState.GROUP_AFTER: 5>
 |      In a group, after required params.
 |
 |
 |  GROUP_BEFORE = <ParamState.GROUP_BEFORE: 2>
 |      In a group, before required params.
 |
 |
 |  LEFT_SQUARE_BEFORE = <ParamState.LEFT_SQUARE_BEFORE: 1>
 |      Left square backets before required params.
 |
 |
 |  OPTIONAL = <ParamState.OPTIONAL: 4>
 |      Positional-or-keyword or positional-only params that now must have default values.
 |
 |
 |  REQUIRED = <ParamState.REQUIRED: 3>
 |      Required params, positional-or-keyword or positional-only (we don't know yet).
 |
 |      Renumber lefft groups!
 |
 |
 |  RIGHT_SQUARE_AFTER = <ParamState.RIGHT_SQUARE_AFTER: 6>
 |      Right square brackets after required params.
 |
 |
 |  START = <ParamState.START: 0>
 |      State we're in before we've seen anything.
 |
 |      Legal transitions: to LEFT_SQUARE_BEFORE or REQUIRED.
```

But maybe that would be overkill? I don't know how often people are calling `help()` in the REPL if they want information about something in `clinic.py`

Reply:
That's a super neat trick, but I'm not sure it makes much sense in Argument Clinic, as you yourself hinted to.

Reply:
Yup, not worth the added complexity here